feat: enable STX by default with migration and notification

httpJunkie opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

This PR enables Smart Transactions (STX) by default through migration number 135 for users who have either opted out or haven't interacted with the STX toggle, provided they have no recorded STX activity. 

Docs: [SmartTransactionsBannerAlert](https://www.notion.so/consensys/SmartTransactionsMigrationBanner-Feature-Documentation-182fc61a326e8076936be5e171b1ec69)

How it works:

- Upon Migration 135, alert displays on transaction confirmations:
 - Legacy transaction flow
 - New transaction flow (experimental)
 - Swaps confirmation flow
 - Contract deployment
 - Contract interactions (minting, etc.)

In the case a user migrates from a previous version of the extension and the migration runs and sets STX toggle "ON" in `Settings > Advanced > Smart Transactions`, they will receive an STX Banner Alert on transaction confirmation screens until dismissed through a close button, or by clicking on the "Higher success rates" link within the alert that goes to: [What is 'Smart Transactions'?](https://support.metamask.io/transactions-and-gas/transactions/smart-transactions/) for more information.

Edge Cases: 

If a user is new and setting up a wallet for the first time, they will not receive the Banner Alert. If a user imports a new wallet during a fresh install of the extension on a new browser or recovers a wallet, it's possible they may not see the alert if STX was on in a previous install. The STX Banner Alert is dismissed and will not show again if a user is in the state to get shown the banner and toggles STX off independently even if they do not physically dismiss the STX Banner Alert.

Migration Logic:

1. If `smartTransactionsOptInStatus` is `null` (new/never interacted)
 -  Sets status to true
 - Enables notification flag
2. If status is false (previously opted out):
 - With no Ethereum Mainnet STX activity: Sets to true with notification
 - With existing Mainnet STX activity: Preserves user preference
3. If status is true: No changes needed

UI Components:

- Implements SmartTransactionsBannerAlert component for user notification

The notification system bridges the migration changes with the UI, ensuring users are informed of the STX enablement while maintaining their ability to opt out through settings.

**Target release:** [release-12.11.0](https://github.com/MetaMask/metamask-extension/labels/release-12.11.0)
**Affected user base:** ~5.7M users who previously opted out of STX but have no STX activity.

---

## **Related issues**

Fixes: N/A

---

## **Running Unit Tests**
Migration Test: 

```bash
yarn jest app/scripts/migrations/135.test.ts --verbose
```

Smart Transaction Banner Component Test: 

```bash
yarn jest ui/pages/confirmations/components/smart-transactions-banner-alert/smart-transactions-banner-alert.test.ts --coverage=false
```

Confirm Transaction base Test: 

```bash
yarn jest ui/pages/confirmations/confirm-transaction-base/confirm-transaction-base.test.js --coverage=false
```

Preferences Controller Test: 

```bash
yarn jest app/scripts/controllers/preferences-controller.test.ts --coverage=false
```

Transaction Alerts Component Test: 

```bash
yarn jest ui/pages/confirmations/components/transaction-alerts/transaction-alerts.test.js --coverage=false
```

## **Manual testing steps**

**Test Migration (using a wallet/account with no STX Transactions)**

1. Switch branch: `git checkout tags/v12.5.0` and run:
```bash
yarn
yarn webpack
```
 - Generate a `dist/chrome` directory
2. In Chrome Extension Manager, "Load Unpacked"  from this directory
3. Import or setup a wallet without STX transactions, launch the wallet and choose "No Thanks" on the "Enhanced Transaction Protection" popup.
4. Check that toggle is OFF in: `Settings > Advanced > Smart Transactions`
5. Close MetaMask Extension and toggle the Extension "OFF" in the Extension Manager
6. Switch to branch from this PR `git checkout feat/enable-stx-migration`
```bash
yarn
yarn webpack
```
7. Open MetaMask Extension and Check that toggle is ON in: `Settings > Advanced > Smart Transactions`

**Test STX Banner Alert that it shows on Transaction Confirmations and not Sign Confirmations)**
**_(using new confirmations flow)_**

8. Check that `Improved transaction requests` is ON in `Settings > Experimental`
9. Open the E2E TestDapp, try several Signs (ETH Sign, Personal Sign, Sign Typed Data, etc..) and ensure the STX Banner Alert does not show on those confirmations screens.
10. Create a Send transaction to your own wallet for `0.0001` ETH
11. Ensure that Smart Transactions Banner Alert IS showing
12. Start a Swaps transaction on Ethereum Mainnet
13 Ensure that Smart Transactions Banner Alert IS showing


**Test STX Banner Alert that it shows on Transaction Confirmations and not Sign Confirmations)**
**_(using old confirmations flow)_**

14. Check that `Improved transaction requests` is OFF in `Settings > Experimental`
15. Open the E2E TestDapp, try several Signs (ETH Sign, Personal Sign, Sign Typed Data, etc..) and ensure the STX Banner Alert does not show on those confirmations screens.
16. Create a Send transaction to your own wallet for `0.0001` ETH
17. Ensure that Smart Transactions Banner Alert IS showing
18. Start a Swaps transaction on Ethereum Mainnet
19. Ensure that Smart Transactions Banner Alert IS showing
20. Without clicking on Check that "Higher success rates" link (inspect) goes to: [What is 'Smart Transactions'?](https://support.metamask.io/transactions-and-gas/transactions/smart-transactions/)
21. Dismiss the Smart Transactions Banner Alert and set `Improved transaction requests` back to ON in `Settings > Experimental`
22. Create a Send transaction to your own wallet for `0.0001` ETH
23. Ensure that Smart Transactions Banner Alert IS NOT showing

**Congrats, you have manually tested the happy path, now we just need to test the edge cases:**

1. Remove the extensions from Extension Manager and Repeat steps 1 to 10 above to run migration again
2. Create a Send transaction to your own wallet for `0.0001` ETH
3. Ensure that Smart Transactions Banner Alert IS showing
 - DO NOT DISMISS THE ALERT. Instead
4. Open MetaMask Extension and Check that toggle is ON in: `Settings > Advanced > Smart Transactions`
5. Turn it off
6. Create a Send transaction to your own wallet for `0.0001` ETH
7. Ensure that Smart Transactions Banner Alert IS NOT showing
8. Perform any other Signing and/or Transaction Confirmations and ensure there are no modals that show errors and that the Banner Alert does not show anymore.

**Test edge case that after STX migration runs and Banner is being shown that clicking on "Higher success rates" link:**

1. Remove the extensions from Extension Manager and Repeat steps 1 to 10 above to run migration again
2. Create a Send transaction to your own wallet for `0.0001` ETH
3. Ensure that Smart Transactions Banner Alert IS showing
 - DO NOT DISMISS THE ALERT WITH CLOSE BUTTON... INSTEAD
4. Click on "Higher success rates" link and ensure that it goes to: [What is 'Smart Transactions'?](https://support.metamask.io/transactions-and-gas/transactions/smart-transactions/)
5. Open MetaMask Extension and Check that toggle is ON in: `Settings > Advanced > Smart Transactions`
6. Turn it off
7. Create a Send transaction to your own wallet for `0.0001` ETH
8. Ensure that Smart Transactions Banner Alert IS NOT showing
9. Perform any other Signing and/or Transaction Confirmations and ensure there are no modals that show errors and that the Banner Alert does not show anymore.

Because the NEW confirmation flow does not support alerts using hooks that are dismissible, we have used the old style Banner Alert, and it is normal for their to be some variation on where the alert shows up and it's surroundings. But overall they should look similar.

---

## **Screenshots/Recordings**

### **Before**

<img width="150" alt="01-stx_before" src="https://github.com/user-attachments/assets/31c0bd4e-1ce5-4bfc-88bc-70a3c1a8898a" />
<img width="150" alt="02-legacySend_before" src="https://github.com/user-attachments/assets/111662b8-182f-44ce-8a3e-e308e08e3c3f" />
<img width="150" alt="03-legacySwap_before" src="https://github.com/user-attachments/assets/25dafb19-bd63-4a29-b0bc-ac394b8f09cf" />
<img width="150" alt="04-signTypedDataV4_before" src="https://github.com/user-attachments/assets/c512ad84-d921-4b1e-8443-6d10d527ab67" />
<img width="150" alt="05-contractDeployment_before" src="https://github.com/user-attachments/assets/31624828-5e3d-4fc8-8264-64703c6a4087" />
<img width="150" alt="06-contractInteraction_before" src="https://github.com/user-attachments/assets/c132c7ff-64dc-4720-9209-331dbaf2bfe3" />
<img width="200" alt="07-wideSwap_before" src="https://github.com/user-attachments/assets/521e8129-368f-45cc-acbe-9c4f90eeb9ec" />

### **After**

<img width="150" alt="01-stx_after" src="https://github.com/user-attachments/assets/ce821edf-bb86-46af-97aa-fb816208c344" />
<img width="150" alt="02-legacySend_after" src="https://github.com/user-attachments/assets/98f3588c-3557-4f4f-9a10-969e27b14dd4" />
<img width="150" alt="03-legacySwap_after" src="https://github.com/user-attachments/assets/b60c91e8-63af-4d45-947d-d250f39fdb74" />
<img width="150" alt="04-signTypedDataV4_after" src="https://github.com/user-attachments/assets/a3600ed9-17eb-4bf2-b5a9-409857fe1911" />
<img width="150" alt="05-contractDeployment_after" src="https://github.com/user-attachments/assets/b6cf1e7e-6dba-4eae-8e78-ce16d12f7306" />
<img width="150" alt="06-contractInteraction_after" src="https://github.com/user-attachments/assets/d65e6a93-c6ca-4328-9622-089ab5fe5d97" />
<img width="200" alt="07-wideSwap_after" src="https://github.com/user-attachments/assets/2161a3a8-6a05-4b9b-a470-e004900c9fc1" />

---

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests for the new behavior, covering:
  - [x] Version update handling
  - [x] All logic branches:
    - [x] `null` opt-in status
    - [x] `false` opt-in status with no STX activity
    - [x] `false` opt-in status with existing STX activity
    - [x] `true` opt-in status
  - [x] Notification flag setting
  - [x] Error handling
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format where applicable.
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

---

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pulled and built the branch, ran the app, and tested the changes described above).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket and includes the necessary testing evidence (e.g., recordings, screenshots, or detailed descriptions).

httpJunkie committed:
Add migration and tests for enabling STX by default, update preferences-controller with `smartTransactionsOptInStatus` in `controllerMetadata`

httpJunkie committed:
add logging to trace state and metadata change in preferences controller and migration's state transformation.

httpJunkie committed:
add logging in setPreference in preferences controller and actions.ts

httpJunkie committed:
update `VersionedData` in the migration to have preferences:

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Add UI for legacy transaction flow, creation of STX Banner Alert for transaction flows that occur when experimental/Improved transaction requests is not toggled on.

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie commented:
I have read the CLA Document and I hereby sign the CLA

httpJunkie commented:
To test the isolated migration (#135 Only) run:
```bash
yarn jest app/scripts/migrations/135.test.ts --verbose
```
![image](https://github.com/user-attachments/assets/55ac40cf-08a7-4020-8de0-fe2c137f8137)
****

To test the `STXBannerAlert` Component run:
```bash
yarn jest --testPathPattern="ui/pages/confirmations/components/stx-banner-alert/stx-banner-alert.test.js" --coverage=false
```
![image](https://github.com/user-attachments/assets/6b1bcb80-3835-4db9-8c2f-af5c38fd088a)

dan437 commented on app/_locales/en/messages.json:
```diff
@@ -6815,5 +6815,11 @@
+  "smartTransactionsEnabledMessage": {
+    "message": "You're now protected with MetaMask Smart Transactions"
+  },
+  "smartTransactionsLearnMore": {
```
This one is not needed, we already have the "learnMore" one.

httpJunkie replied:
Good call, look at all of these:
![image](https://github.com/user-attachments/assets/7318d24a-03ea-4378-b6a5-1a74f7d953b4)

dan437 commented on app/_locales/en/messages.json:
```diff
@@ -6815,5 +6815,11 @@
   "zeroGasPriceOnSpeedUpError": {
     "message": "Zero gas price on speed up"
+  },
+  "smartTransactionsEnabledMessage": {
```
nit: All keys in this file are for messages, so I don't think we need the "Message" postfix in the key name. We can just say: `smartTransactionsEnabled`

dan437 commented on app/scripts/controllers/preferences-controller.ts:
```diff
@@ -925,11 +935,14 @@ export class PreferencesController extends BaseController<
     preference: keyof Preferences,
     value: Preferences[typeof preference],
   ): Preferences {
+    console.log('Setting preference:', { preference, value });
```
Once the PR is not in Draft anymore, can you please remove all the console.log calls you added from all files?

dan437 commented on app/scripts/migrations/135.test.ts:
```diff
@@ -0,0 +1,131 @@
+
+    const newStorage = await migrate(oldStorage);
+    expect(
+      newStorage.data.PreferencesController?.smartTransactionsOptInStatus,
```
The `PreferencesController` will always be there in this case, so we don't need `?`. It can be: `newStorage.data.PreferencesController.smartTransactionsOptInStatus`. The same in other tests.

httpJunkie replied:
According to their type def:
```ts
export type VersionedData = {
  meta: {
    version: number;
  };
  data: {
    PreferencesController?: {  // Note the ? here
      preferences?: {
        smartTransactionsOptInStatus?: boolean | null;
      };
      smartTransactionsOptInStatus?: boolean | null;
    };
  };
};
```

TypeScript is correctly warning us that PreferencesController might be undefined according to the type definition. (If I don't use it I get a warning)
```
'newStorage.data.PreferencesController' is possibly 'undefined'.ts(18048)
```

Therefore, I do believe the safest thing to do is to keep the optional chaining. LMKWYT...

dan437 commented on app/scripts/migrations/135.test.ts:
```diff
@@ -0,0 +1,131 @@
+        },
+        SmartTransactionsController: {
+          smartTransactionsState: {
+            smartTransactions: {},
```
You can update this test to check if there are existing smart transactions only on Ethereum Mainnet. E.g. this should set it to true:
```
smartTransactions: {
  '0x1': [],
  '0xAA36A7': [mockSmartTransaction], // Sepolia
},
```

httpJunkie replied:
I have made this change ensure that the STX opt-in status is only affected by transactions on Ethereum Mainnet, ignoring any testnet activity. I have also ensured that the migration was changed to ensure this test will pass.

dan437 commented on app/scripts/migrations/135.ts:
```diff
@@ -0,0 +1,101 @@
+  data: {
+    PreferencesController?: {
+      preferences?: {
+        smartTransactionsOptInStatus?: boolean | null;
```
Why is the `smartTransactionsOptInStatus` line here twice?

httpJunkie replied:
Thanks for the heads up, I will most likely be removing the root level property in the `controllerMetadata` inside `preferences-controller` and that should fix this. I was initially trying to figure out the right place for it and this is a valid point.

dan437 commented on app/scripts/migrations/135.ts:
```diff
@@ -0,0 +1,101 @@
+export const version = 135;
+
+function transformState(state: VersionedData['data']) {
+  console.log('Migration 135 state:', JSON.stringify(state, null, 2));
```
Just a reminder to remove all console.log calls in this file as well.

dan437 commented on shared/constants/alerts.ts:
```diff
@@ -2,6 +2,7 @@ export enum AlertTypes {
   unconnectedAccount = 'unconnectedAccount',
   web3ShimUsage = 'web3ShimUsage',
   invalidCustomNetwork = 'invalidCustomNetwork',
+  stxMigration = 'stxMigration',
```
With new code for smart transactions we aim to not use the `stx` shortcut in naming and use the full name instead (e.g. `smartTransactions` or `smartTransaction`) to make code clearer. Can you please do a little renaming everywhere you used the word `stx`, including file names? Thank you

dan437 commented on ui/pages/confirmations/components/stx-banner-alert/stx-banner-alert.js:
```diff
@@ -0,0 +1,45 @@
+
+const STXBannerAlert = () => {
+  const dispatch = useDispatch();
+  const shouldShow = useSelector(stxAlertIsOpen);
```
Can you rename `stxAlertIsOpen` to `shouldShowSmartTransactionsMigrationAlert` to make it little clearer?
```
 const showSmartTransactionsMigrationAlert = useSelector(shouldShowSmartTransactionsMigrationAlert);
```

Alternatively, you can also do this renaming:
```
 const smartTransactionsMigrationAlertSeen = useSelector(wasSmartTransactionsMigrationAlertSeen);

if (smartTransactionsMigrationAlertSeen) {
    return null;
  }
```

dan437 commented on ui/ducks/alerts/stx-migration.js:
```diff
@@ -0,0 +1,34 @@
+    showSTXMigrationAlert: (state) => {
+      state.state = ALERT_STATE.OPEN;
+    },
+    dismissSTXMigrationAlert: (state) => {
```
nit: I would change the name to `dismissSmartTransactionsMigrationAlert`

dan437 commented on ui/pages/confirmations/components/transaction-alerts/transaction-alerts.js:
```diff
@@ -81,6 +82,7 @@ const TransactionAlerts = ({
   return (
     <div className="transaction-alerts">
       <BlockaidBannerAlert txData={txData} />
+      <STXBannerAlert />
```
Once possible, can you please add a screenshot with this in the PR description?

httpJunkie replied:
The comment below this one shows the UX from before and after migration with settings and transaction screen & alert before and after dismissal.

httpJunkie committed:
update `stx-migration.js` to include `extraReducers`, added test and passing, update console.logs to debug in migration code for testing manually

httpJunkie committed:
Add logs and alet to ensure component is mounting and state is as intended

httpJunkie committed:
remove alert and update logs

httpJunkie committed:
registering our stxMigration reducer in ducks index

httpJunkie committed:
Remove logs from migration script; update stx-migration logs to debug alert dismissal issue.

httpJunkie committed:
update stx-migration in ducks to dismissAndDIsable alert state once the user has seen our message and has dismissed.

httpJunkie committed:
revert changes to wrong file

httpJunkie committed:
update stx-migration ducks file to consider dismiss and disable alert.

Update stx-migration ducks test.

httpJunkie committed:
update path to store in test file and  directly tests the thunk behavior without complex store mocking.

httpJunkie committed:
add logs to check extraReducer logic and state.

httpJunkie commented:
**_User Experience for Migration from 12.5.0 => 12.7.2 for a User with no existing STX Transactions_**

Logging into wallet before running a Migration:
![image](https://github.com/user-attachments/assets/a316533b-9a2b-434e-aab1-2dfc7bd2e22c)

Logging into wallet after running a Migration:
![image](https://github.com/user-attachments/assets/69962887-54e0-4b6a-9629-38a08363bf6c)

First Transaction after running Migration:
![image](https://github.com/user-attachments/assets/55e4069f-e0c8-4602-b79d-73afe3f58827)

Second Transaction after dismissing the first alert and rejecting transaction and starting a new Send Transaction:
![image](https://github.com/user-attachments/assets/76d3aa03-bd61-46d8-84b5-1a62dc55374c)

httpJunkie committed:
Add permanent alert dismissal for MetaMask Smart Transactions onboarding message. When a user dismisses the alert (via close or "Learn more"), their preference is stored and persists across transaction sessions. The dismissal is handled through the alerts state system and uses the existing setAlertEnabledness functionality. Remove comments.

httpJunkie committed:
With changes to permanent alert dismissal, we needed to update our tests for the STXBannerAlert and TransactionAlerts tests.

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
First stab at trying to implement alert for new confirmation/transaction flow. Does not let me close alert or show link.

httpJunkie committed:
added an action to dismiss

httpJunkie committed:
feat: add STX migration toast to new confirmation flow

httpJunkie committed:
move toast to top of page and try to default to darkmode

danjm committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
chore: cleanup STX migration implementation

Removes console logs, unused toast components, and redundant preference config.

httpJunkie committed:
Merge branch 'feat/enable-stx-migration' of github.com:MetaMask/metamask-extension into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Add Banner Alert to the Swaps page (top)

httpJunkie committed:
chore: remove 'smartTransactionsLearnMore' for use of 'learnMoreUpperCase' which already exists.

httpJunkie committed:
Change `smartTransactionsLearnMore` to 'learnMore' on STXBannerAlert

httpJunkie committed:
fix: Add 'transaction-alerts' style to wrap component as it's not made for the new transaction screen.

httpJunkie committed:
chore: remove 'Message" postfix

httpJunkie committed:
fix: This change ensures that the STX opt-in status is only affected by transactions on Ethereum Mainnet, ignoring any testnet activity.

httpJunkie committed:
chore: Rename any and all use of word: 'stx or STX' for variables and files within this PR

dan437 commented on app/scripts/migrations/135.ts:
```diff
@@ -0,0 +1,86 @@
+    state.SmartTransactionsController.smartTransactionsState;
+
+  // Only check mainnet transactions
+  const mainnetTransactions = smartTransactions['0x1'] || [];
```
We should be using constants for chainIds. You can do this here and in other places where you used `0x1`:
```
import { CHAIN_IDS } from '@metamask/transaction-controller';
const mainnetTransactions = smartTransactions[CHAIN_IDS.MAINNET] || [];
```

dan437 commented on ui/helpers/constants/zendesk-url.js:
```diff
@@ -49,6 +49,8 @@ const ZENDESK_URLS = {
     'https://support.metamask.io/networks-and-sidechains/managing-networks/verifying-custom-network-information/',
   NETWORK_DEPRECATED:
     'https://support.metamask.io/networks-and-sidechains/eth-on-testnets/',
+  SMART_TRANSACTIONS_LEARN_MORE:
```
We already have this constant, can you use that one? https://github.com/MetaMask/metamask-extension/blob/main/shared/constants/smartTransactions.ts#L30

There are 2 reasons why it's in `shared/constants/smartTransactions.ts`:
- To keep smart transactions constants in one place
- It shouldn't be in zendesk-url.js since it's not a zendesk URL

httpJunkie replied:
Yeah this was an oversight. Thanks for the heads up

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
chore: address issues in PR (use constants), remove zendesk url, remove margin from transaction alert in new confirmation flow (too much margin)

httpJunkie committed:
fix: Update SmartTransactionsBannerAlert to encompass styling needed to adjust margins for use in various components. Styling modifications are confined to our component code and I've added a `smart-transactions-banner-alert.d.ts` file and JSDocs for our SmartTransactionsBannerAlert component.

httpJunkie committed:
fix: update the index file for SmartTransactionsBannerAlert as we are no longer exporting a default.

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Update Smart Transactions Migration redux/ducks test to ensure we are not printing console errors in the terminal output using `consoleErrorSpy`. Added spy verification to ensure error is logged with correct message and error object.

httpJunkie committed:
fix: test needed to be updated after changing naming in `SmartTransactionBannerAlert` and `learnMore` link/URL constant change.

httpJunkie committed:
chore: run prettier fix on `messages.json`

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
fix: fitness function failure, convert our JS files to TypeScript since project rules against adding new JS/JSX file.

httpJunkie committed:
fix: fitness function failure, convert our JS files to TypeScript

httpJunkie committed:
test: update snapshots to include STX banner container

httpJunkie committed:
fix: remove unused d.ts file no longer needed as SmartTransactionsBannerAlert is now written in TS, also fix lint error on stx-migration ducks test.

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
fix: These are old tests planned for disable. The quicker way to solve this is to disable them with your PR. dBrochetto

httpJunkie committed:
Merge branch 'feat/enable-stx-migration' of github.com:MetaMask/metamask-extension into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

matthewwalsh0 commented on test/e2e/tests/swaps/swap-eth.spec.ts:
```diff
@@ -9,7 +9,7 @@ import {
 } from './shared';
 
-describe('Swap Eth for another Token @no-mmi', function () {
+xdescribe('Swap Eth for another Token @no-mmi', function () {
```
Accidental?

httpJunkie replied:
I can revert.

matthewwalsh0 commented on test/e2e/tests/metrics/swaps.spec.js:
```diff
@@ -90,7 +90,7 @@ async function mockSegmentAndMetaswapRequests(mockServer) {
 }
 
-describe('Swap Eth for another Token @no-mmi', function () {
+xdescribe('Swap Eth for another Token @no-mmi', function () {
```
Accidental?

matthewwalsh0 commented on ui/ducks/alerts/smart-transactions-migration.ts:
These reducers seem to be legacy code as they haven't been used in a long time, plus this is a huge amount of code to ultimately persist a single boolean to track whether an alert was dismissed.

Could we remove all this, and just have:

- `useSelector` within the `SmartTransactionsBannerAlert` that references the `alertEnabledness` state.
- `dispatch` to `setAlertEnabledness` action in the `onClose` handler

Would that achieve the same with two or three lines of code?

httpJunkie replied:
TBH, I kind of got stuck in trying to figure our what was being used and what wasn't for state. I probably piled into this idea a little too much without thinking of a more optimal solution. Change made. Great point. Reduces some extra code that would need to be cleaned up when we remove these.

matthewwalsh0 commented on ui/pages/confirmations/components/transaction-alerts/transaction-alerts.test.js:
```diff
@@ -558,3 +563,47 @@ describe('TransactionAlerts', () => {
 });
+
+describe('Smart Transactions Migration Alert', () => {
+  it('should show smart transactions banner when shouldShowSmartTransactionsMigrationAlert is true', () => {
```
Minor, we're referencing the `shouldShowSmartTransactionsMigrationAlert` selector in the test names but in the context of the test we don't know anything about that as we're just setting state.

For example, could this just be `shows if alert not closed in state`?

httpJunkie replied:
I have updated this test. Thanks this is good feedback.

httpJunkie commented on ui/pages/confirmations/components/smart-transactions-banner-alert/smart-transactions-banner-alert.tsx:
```diff
@@ -0,0 +1,71 @@
+      <BannerAlert
+        severity={BannerAlertSeverity.Info}
+        onClose={() => {
+          console.log('Dismiss clicked');
```
Accidental, being removed.

httpJunkie committed:
fix: removed console.log accidentally left in.

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
This seems to be the correct way to skip these tests rather than `xdescribe`

httpJunkie committed:
Merge branch 'feat/enable-stx-migration' of github.com:MetaMask/metamask-extension into feat/enable-stx-migration

httpJunkie committed:
refactor: simplify smart transactions migration alert

Removed legacy alert reducers in favor of direct state access for smart transactions banner

httpJunkie committed:
fix: let's make the test descriptions more accurate to what we're actually testing

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
fix: update snapshot to include smartTransactionsOptInStatus

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'feat/enable-stx-migration' of github.com:MetaMask/metamask-extension into feat/enable-stx-migration

httpJunkie committed:
added snapshots for Banner Alert

georgeweiler committed:
chore: updates snapshots for the connect-account-modal test

georgeweiler committed:
Merge branch 'main' into feat/enable-stx-migration

georgeweiler committed:
Merge branch 'feat/enable-stx-migration' of github.com:MetaMask/metamask-extension into feat/enable-stx-migration

georgeweiler committed:
Revert "chore: updates snapshots for the connect-account-modal test"

This reverts commit b2658939a7f31bd1e78c8d310b087093434762c6.

httpJunkie committed:
fix: revert to ^HEAD on problematic test.

httpJunkie committed:
Merge branch 'feat/enable-stx-migration' of github.com:MetaMask/metamask-extension into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
disable STX for these tests:  adding the baseFixtureOptions with smartTransactionsOptInStatus: false and using it in all tests

httpJunkie committed:
Merge branch 'feat/enable-stx-migration' of github.com:MetaMask/metamask-extension into feat/enable-stx-migration

httpJunkie committed:
match the structure where metamask should ibe nested under data in our fixtures object.

httpJunkie committed:
revert changes to `swaps-notifications.spec.ts` as test results did not improve on CI.

httpJunkie committed:
Add logic to determine confirmationType in Wrapper

httpJunkie committed:
chore: Move all logic for conditionally rendering the SmartTransactionBanner Alert into itself.

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
chore: update SmartTransactionsBannerAlert to maintain backward compatibility with the existing test.

httpJunkie committed:
fix: update smart-transaction-banner-alert test so it verifies our new context-aware functionality and keeps existing behavior. testing edge case: banner doesn't show when it shouldn't. test margin types systematically, and verifying that the banner handles being outside of ConfirmContext appropriately

fix: update smart-transaction-banner-alert modified the shouldRender logic to handle no context differently.

httpJunkie committed:
fix: auto-dismiss STX migration banner when user disables STX

Add automatic dismissal of Smart Transactions migration banner when
user manually disables Smart Transactions from settings, preventing
stale or incorrect banner state.

httpJunkie committed:
fix: handle STX banner dismiss when toggling STX off

Automatically dismisses Smart Transactions migration banner when user manually
disables Smart Transactions to prevent stale UI state.

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
fix: linting / TS errors

httpJunkie committed:
fix: update snapshots for `snap-ui-address`, `connect-accounts-modal`, and `confirm`
update `smart-transactions-banner-alert` (remove `as string`) to ensure confirm test passes
update `confirm-transaction-base.test` to mock background connection for `setAlertEnabledness`

httpJunkie commented:
Test results as of Fri Dec 13th 2024 4pm:
```bash
yarn jest --silent --verbose=false
```
![image](https://github.com/user-attachments/assets/9bed3ae4-2659-49e6-917d-e2e3d05b279e)

httpJunkie committed:
fix: add `as string` back to satisfy linting

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
fix: update snapshot for connect-account-modal to have the same translate value for CI

httpJunkie committed:
fix: update snapshot to match CI transform for `snap-ui-address`

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
fix: update STX Banner Alert text to match new design mockups, ensure that text and links are as mockup shows, update test for STX Banner Alert component.

httpJunkie committed:
fix: improve STX banner display logic for new installs

Add handling of Smart Transactions banner for fresh installs vs migrated users:

- Add smartTransactionsMigrationApplied flag to preferences
- Update migration 135 to set migration flag
- Update SmartTransactionsBannerAlert to check migration status
- Update tests to handle migration flag

The banner will now show ONLY for users who have been migrated to STX,
not for fresh installs.

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
fix: update preferences-controller test because we added smartTransactionsMigrationApplied but didn't update the test expectations

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
fix: update snapshot for `confirm-send-ether.test.js`

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
fix: add test for prepare-swap-page.test.js

httpJunkie committed:
fix: updated snapshots from failing tests.

httpJunkie committed:
fix: snapshot update for connect-accounts-modal

httpJunkie committed:
fix: update snapshot for `snap-ui-address.test.tsx`

httpJunkie committed:
fix: add `smartTransactionsMigrationApplied` to `expectedMissingState`

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
fix: try adding withPreferencesController with smartTransactionsMigrationApplied added in error.spec.js

httpJunkie committed:
fix: fix formatting on error.spec.js

httpJunkie committed:
fix: reset `errors.spec.js` to version on main

httpJunkie committed:
fix: remove comments in preference-controller

httpJunkie committed:
feat: update how we check for transaction confirmations

httpJunkie committed:
fix: update `confirm.test.tsx` and snapshot.

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

dan437 committed:
Update snapshots for E2E tests

Linting

dan437 commented on ui/pages/confirmations/components/smart-transactions-banner-alert/smart-transactions-banner-alert.tsx:
```diff
@@ -0,0 +1,148 @@
+      setAlertEnabledness(AlertTypes.smartTransactionsMigration, false);
+    }, [dispatch]);
+
+    const shouldRender =
```
Can you simplify this code, e.g. like this:
```
const alertConditions = alertEnabled && smartTransactionsOptIn && smartTransactionsMigrationApplied;

const shouldRender = currentConfirmation === null
  ? alertConditions
  : alertConditions && isCorrectDeveloperTransactionType(currentConfirmation?.type as TransactionType);

```

dan437 commented on ui/pages/confirmations/components/transaction-alerts/transaction-alerts.test.js:
```diff
@@ -558,3 +568,65 @@ describe('TransactionAlerts', () => {
+          networkConfigurationsByChainId: {
+            [CHAIN_ID_MOCK]: {
+              chainId: CHAIN_ID_MOCK,
+              // other required network properties
```
You can remove this comment.

httpJunkie replied:
will do, I will check for others too.

dan437 committed:
Update a Jest snapshot

dan437 committed:
Update an E2E test

httpJunkie committed:
fix: remove root level duplicated `smartTransactionsOptInStatus` (we have already removed in `preference-controller` and I forgot to remove from the migration. Removed unneeded comments.

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
fix: cleaned up test for migration after removing `smartTransactionsOptInStatus` from root level of PreferencesController.

Also, cleaned up migration test for readability.

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
fix: remove root level `smartTransactionsOptInStatus` from snapshots.

httpJunkie committed:
Merge branch 'feat/enable-stx-migration' of github.com:MetaMask/metamask-extension into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
fix: Remove redundant test for an impossible state (alert enabled before or without migration) as conditions are already tested individually.

httpJunkie committed:
refactor: streamline SmartTransactionsBannerAlert logic

- Replace inline Redux selector with `getSmartTransactionsOptInStatusInternal`
- Add `getSmartTransactionsMigrationAppliedInternal` for consistent state handling
- Centralize banner visibility logic in `alertConditions` for clarity
- Consolidate alert dismissal logic in `dismissAlert` and remove redundant `dispatch`
- Simplify state typing and remove unused types
- Merge duplicate dismissal tests for shared functionality

Alert state is now managed via AlertController's `setAlertEnabledness`. Selectors are reusable, logic is more maintainable, and tests better reflect component behavior.

httpJunkie committed:
refactor: simplify logic in migration 135 and fix formatting in smart-transactions-alert-banner.test.tsx

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
fix: remove skipped tests

httpJunkie committed:
fix: close SmartTransactionsMigrationNotification on `swap-eth.spec.ts`

httpJunkie committed:
fix: Scroll "View all quotes" button into view before clicking in STX migration test

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Remove `"bridge.dev-api.cx.metamask.io",` from my branch to mirror `main`

httpJunkie committed:
Merge branch 'feat/enable-stx-migration' of github.com:MetaMask/metamask-extension into feat/enable-stx-migration

dan437 commented on ui/pages/confirmations/components/smart-transactions-banner-alert/smart-transactions-banner-alert.tsx:
```diff
@@ -0,0 +1,124 @@
+    );
+
+    const smartTransactionsMigrationApplied = useSelector(
+      getSmartTransactionsMigrationAppliedInternal,
```
nit: can we remove the `Internal` postfix? So it would be: `getSmartTransactionsMigrationApplied`

forest-diggs-consensys commented:
Let's make sure to test upgrading from several older versions as well
